### ISSUE TITLE
Refactor CAPV controller context

### ISSUE THREAD

**Issue by zhanggbj** [kubernetes-sigs/cluster-api-provider-vsphere#2295]
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Currently CAPV has a customized Controller Context which wraps context.Context along with other essential controller-related information as [here](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/pkg/context/controller_manager_context.go#L35). This context is propagated across controllers and reconcilers. However, this approach comes with drawbacks:

- Not align with the recommended practices outlined by the `containedctx` lint, which emphasizes that 'Contexts should not be stored inside a struct type, but instead passed to each function that needs it'.
- Hard to understand the function and may need further investigation for better comprehension about what it needs, for example `func (r clusterReconciler) reconcileNormal(ctx *context.ClusterContext)`
- contains mixed usage of CAPV controller context and golang context, sometimes capv will initialize new golang context to call some API. It is difficult to enable contextual logging and tracing.

Given the sensitivity of context usage, particularly its impact across controllers, propose a refactoring of the controller context, aiming to split it into golang context and the CAPV controller context. This will be achieved incrementally through the following steps:

- [x] Enable the linter `containedctx` and use '//nolint:containedctx' to exclude current findings. Refer to #2058 . (@zhanggbj #2297 )
- [x] rename all "CAPV context's" in a consistent way: e.g. ControllerContext => controllerCtx, also like imported package alias, function signature (@zhanggbj #2309 )
- [x] Introduce new ctx context.Context parameters to controller and reconciler and underlying functions, such as functions under /pkg, also fix Unit Test. We can break this down by controllers as outlined below, but let's begin and be open to adjustments as needed.  
	- [x] controllers/clustermodule_reconciler.go (@sbueringer https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2352)
	- [x] controllers/serviceaccount_controller.go (@sbueringer https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2352)
	- [x] controllers/servicediscovery_controller.go (@sbueringer https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2352)
	- [x] controllers/vspherecluster_controller.go (@zhanggbj https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2339)
	- [x] controllers/vspherecluster_reconciler.go (@zhanggbj https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2339)
	- [x] controllers/vsphereclusteridentity_controller.go (@MrDXY) #2374
	- [x] controllers/vspheredeploymentzone_controller.go (@zhanggbj https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2369)
	- [x] controllers/vspheredeploymentzone_controller_domain.go (@zhanggbj https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2369)
	- [x] controllers/vspheremachine_controller.go (@Ankitasw) #2376
	- [x] controllers/vspherevm_controller.go (@xiujuanx) #2398 
	- [x] controllers/vspherevm_ipaddress_reconciler.go (@xiujuanx) #2398 
	- [x] controllers/vmware/vspherecluster_reconciler.go
	- [x] controller initiation in main.go (https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2339)

- [x] get rid of our containedctx's exception and removed context.Context from capv customized controller Context (@zhanggbj https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/2431)
- [x] check that there are no leftover `context.TODO()`'s (indicates we missed something to refactor above) (@zhanggbj https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/2431)

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by sbueringer** [comment]
Sounds perfect. Thank you!

**Comment by sbueringer** [comment]
> Introduce new ctx context.Context parameters to funcs, also fix Unit Test. We can break this down by controllers as outlined below, but let's begin and be open to adjustments as needed.

Just stating the obvious probably. The idea would be basically to use the `ctx` that controller-runtime hands us over in the Reconcile func and then pass it down to wherever we need it. This would be:
* places where today we use the ctx from our CAPV contexts
* places where we use context.TODO/context.Background (in a lot of cases this is when we make a Get/List/... call via the controller-runtime client)

**Comment by zhanggbj** [comment]
The first PR to pass down context to controller/reconciler are in review https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2339.

Please feel free to pick up other tasks listed above once it is merged. We can assign this issue to multiple assignees :) Thanks!

**Comment by sbueringer** [comment]
Let's hold this for a bit (hopefully only a few days). I started to unblock the logging work and had to add more context handling across half the repo. Let's get this one in first to avoid duplicate work:  #2076

**Comment by sbueringer** [comment]
Updated the tasks above with the controller already refactored via #2352

**Comment by sbueringer** [comment]
#2352 merged

@zhanggbj  I think we can continue here 



cc @Madhur97 @Ankitasw If you have time to pick up some of the tasks here (let us know if you need more context)

**Comment by zhanggbj** [comment]
Also CC @MrDXY @xiujuanx feel free to pick up tasks, this is fundamental work for tracing and logging :-)

**Comment by MrDXY** [comment]
I'll take this one! 😄 

- [ ] controllers/vsphereclusteridentity_controller.go

**Comment by Ankitasw** [comment]
I can pick  `controllers/vspheremachine_controller.go`

**Comment by zhanggbj** [comment]
/assign @Ankitasw 
(trying to assign it to multiple assignees, but it didn't work?)

**Comment by zhanggbj** [comment]
I'll continue with this one.
- [ ] controllers/vspheredeploymentzone_controller.go

**Comment by xiujuanx** [comment]
I would like to help on this one:

- [ ] controllers/vspherevm_controller.go

**Comment by sbueringer** [comment]
Thx folks! I've updated the list above

**Comment by Madhur97** [comment]
I can pick `controllers/vspherevm_ipaddress_reconciler.go`

**Comment by zhanggbj** [comment]
About the logger which may relate to context refactoring for each controller, I think we could leave them for separate PRs to keep our PR easy to review, and logging refactoring are tracking by #2076  
CC @MrDXY @Ankitasw @xiujuanx @Madhur97

**Comment by sbueringer** [comment]
Absolutely. Let's not start using contextual logging in these PRs.

**Comment by Ankitasw** [comment]
@xiujuanx not urgent, but did any of the work started for `controllers/vspherevm_controller.go`? If not I can take that up.

**Comment by chrischdi** [comment]
Re-assigning `controllers/vspherevm_ipaddress_reconciler.go ` to @xiujuanx because its included in #2398 .

**Comment by adityabhatia** [comment]
I believe first 3 entries:  `controllers/clustermodule_reconciler.go`, `controllers/serviceaccount_controller.go`, `controllers/servicediscovery_controller.go` were already fixed in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/commit/09ff688ddb29e6a8900c1aa02241f5b26de15713

**Comment by zhanggbj** [comment]
@adityabhatia yes, just linked them in case of any confusion.

**Comment by chrischdi** [comment]
#2398 should be merging now too.

Afterwards we can do the final steps and verify that we didn't miss a case :-)

**Comment by chrischdi** [comment]
Thanks for taking the next tasks @zhanggbj !

**Comment by Ankitasw** [comment]
I have some bandwidth, can I pick this:
`get rid of our containedctx's exception and removed context.Context from capv customized controller Context`

**Comment by sbueringer** [comment]
Sounds good, thank you!
